src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs

```diff
@@ -390,6 +390,55 @@ private static bool CheckConstraintsSingleType(TypeSymbol type, CheckConstraints
         }
 
         public static bool CheckConstraints(
+            this TupleTypeSymbol tuple,
+            ConversionsBase conversions,
+            SyntaxNode typeSyntax,
+            ImmutableArray<Location> elementLocations,
+            Compilation currentCompilation,
+            DiagnosticBag diagnostics)
+        {
+            NamedTypeSymbol type = tuple.TupleUnderlyingType;
+            if (!RequiresChecking(type))
+            {
+                return true;
+            }
+
+            bool result = !typeSyntax.HasErrors;
+
+            var diagnosticsBuilder = ArrayBuilder<TypeParameterDiagnosticInfo>.GetInstance();
+            var underlyingTupleTypeChain = ArrayBuilder<NamedTypeSymbol>.GetInstance();
+            TupleTypeSymbol.GetUnderlyingTypeChain(type, underlyingTupleTypeChain);
+
+            int offset = 0;
+            foreach (var underlyingTuple in underlyingTupleTypeChain)
+            {
+                ArrayBuilder<TypeParameterDiagnosticInfo> useSiteDiagnosticsBuilder = null;
+                result &= CheckTypeConstraints(underlyingTuple, conversions, currentCompilation, diagnosticsBuilder, ref useSiteDiagnosticsBuilder);
+
+                if (useSiteDiagnosticsBuilder != null)
+                {
+                    diagnosticsBuilder.AddRange(useSiteDiagnosticsBuilder);
+                }
+
+                foreach (var pair in diagnosticsBuilder)
+                {
+                    var ordinal = pair.TypeParameter.Ordinal + offset;
```

Comment:
>What happens if we add a "class" constraint for the last type parameter in ValueTuple`8? 

It will be caught by reference constraint check as `IsReferenceType` is false for the tuple type `(int)`.
```cs
if (typeParameter.HasReferenceTypeConstraint && !typeArgument.IsReferenceType)
```
>What would be the Ordinal for this constraint failure?

It's seven and it will be caught at the first iteration (offset=0).

>Why is it a right thing to associate this failure with any element in this case? 

I didn't quite understand. we iterate over failures and associate a location with the diagnostic.

I'll add a test for this (although, it currently pass without any change).